plugin/cog/cog-rangereader-http/src/main/java/it/geosolutions/imageioimpl/plugins/cog/HttpRangeReader.java

```diff
@@ -107,20 +105,36 @@ public HttpRangeReader(BasicAuthURI uri, int headerLength) {
             return data;
         }
 
-       Instant start = Instant.now();
-       List<AsyncHttpCallback> callbacks = new ArrayList<>(ranges.length);
+        Instant start = Instant.now();
+        List<AsyncHttpCallback> callbacks = new ArrayList<>(ranges.length);
+        Map<Long, byte[]> values = new HashMap<>();
+        int missingRanges[] = new int[ranges.length];
+        int missing = 0;
 
         for (int i = 0; i < ranges.length; i++) {
-            Call call = client.newCall(buildRequest(ranges[i]));
-            AsyncHttpCallback callback = new AsyncHttpCallback().startPosition(ranges[i][0]);
-            call.enqueue(callback);
-            callbacks.add(callback);
+            byte[] dataRange = data.get(ranges[i]);
+            // check for available data
+            if (dataRange == null) {
+                Call call = client.newCall(buildRequest(ranges[i]));
+                AsyncHttpCallback callback = new AsyncHttpCallback().startPosition(ranges[i][0]);
+                call.enqueue(callback);
+                callbacks.add(callback);
+                // Mark the range as missing
+                missingRanges[missing++] = i;
+            } else {
+                values.put(ranges[i][0], dataRange);
+            }
         }
 
         awaitCompletion(callbacks);
         Instant end = Instant.now();
         LOGGER.fine("Time to read all ranges: " + Duration.between(start, end));
-        return data;
+        for (int k = 0; k < missing; k++) {
+            long range = ranges[missingRanges[k]][0];
```

Comment:
I'm afraid this might break... the async calls can take some time, and they can be more than 10. What if one of the ranges just read is stored, and then becomes soft, and expires, before all the othe ranges are read? You'd read a null here.

It would be better if the asynch call read into the local map, which has hard references, and then you'd transfer to the shared cache at the end, instead.

Reply:
Updated